pkg/apiserver/handlers/webhook/convert_crd_test.go

```diff
@@ -0,0 +1,173 @@
+// limitations under the License.
+
+package webhook
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var (
+	unstructuredTestData = map[string]interface{}{
+		"kind":       "kind",
+		"apiVersion": "apiVersion",
+	}
+	invalidUnstructuredTestData = map[string]interface{}{
+		"test": "test",
+	}
+
+	rawUnstructured, _ = json.Marshal(unstructuredTestData)
+	invalidRawData, _  = json.Marshal(invalidUnstructuredTestData)
+
+	testConvertFunc = func(Object *unstructured.Unstructured, version string) (*unstructured.Unstructured, metav1.Status) {
+		return &unstructured.Unstructured{}, metav1.Status{
+			Status: metav1.StatusSuccess,
+		}
+	}
+)
+
+func TestHandleCRDConversion(t *testing.T) {
+	testCases := []struct {
+		name               string
+		contentType        string
+		requestBody        interface{}
+		expectedStatusCode int
+	}{
+		{
+			name:               "invalid content type",
+			contentType:        contentTypeHtml,
+			expectedStatusCode: http.StatusUnsupportedMediaType,
+		},
+		{
+			name:               "invalid request body",
+			contentType:        contentTypeJson,
+			expectedStatusCode: http.StatusBadRequest,
+		},
+		{
+			name:               "unsupported GVK",
+			contentType:        contentTypeJson,
+			expectedStatusCode: http.StatusBadRequest,
+			requestBody: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"kind":       "APIVersions",
+					"apiVersion": "v1",
+				},
+			},
+		},
+		{
+			name:               "convert successfully",
+			contentType:        contentTypeJson,
+			expectedStatusCode: 0,
+			requestBody: &v1.ConversionReview{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ConversionReview",
+					APIVersion: "apiextensions.k8s.io/v1",
+				},
+				Request: &v1.ConversionRequest{
+					UID:               "uid",
+					DesiredAPIVersion: "v1",
+					Objects: []runtime.RawExtension{
+						{
+							Raw: invalidRawData,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:               "convert failed",
+			contentType:        contentTypeJson,
+			expectedStatusCode: 0,
+			requestBody: &v1.ConversionReview{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ConversionReview",
+					APIVersion: "apiextensions.k8s.io/v1",
+				},
+				Request: &v1.ConversionRequest{
+					UID:               "uid",
+					DesiredAPIVersion: "v1",
+					Objects: []runtime.RawExtension{
+						{
+							Raw: rawUnstructured,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	fn := HandleCRDConversion(testConvertFunc)
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			resp := NewResponseWriter()
+			b, _ := json.Marshal(tt.requestBody)
+			request, _ := http.NewRequest("", "", bytes.NewReader(b))
```

Comment:
we could use https://pkg.go.dev/net/http/httptest#NewRequest instead, for a "server-side" request object? It will also panic on case of error.

Reply:
Chaned to `httptest.NewRequest".